src/audio/module_adapter/module/waves/waves.c

```diff
@@ -492,37 +536,62 @@ static int waves_effect_message(struct processing_module *mod, void *data, uint3
 	return 0;
 }
 
-/* apply codec config */
-static int waves_effect_config(struct processing_module *mod)
+/* apply config blob */
+static int waves_effect_apply_config_blob_from_cache(struct processing_module *mod)
 {
 	struct comp_dev *dev = mod->dev;
 	struct module_data *codec = &mod->priv;
 	struct waves_codec_data *waves_codec = codec->private;
+
+	comp_info(dev, "waves_effect_apply_config_blob_from_cache()");
+
+	if (waves_codec->config_blob) {
+		return waves_effect_message(mod, waves_codec->config_blob,
+				waves_codec->config_blob_size);
+	}
+	return 0;
+}
+
+static int waves_effect_handle_param_message(struct processing_module *mod,
```

Comment:
How would the waves library handle the case that client calls MaxxEffect_Message() during processing? For example, the normal case:
```
prepare()               process() *n                   reset()
MaxxEffect_Message() -> MaxxEffect_Process() *n ... -> MaxxEffect_Reset()
```
the specific case that the config setting is interrupted during the process sequence (the interrupted config is valid and could be either equal to the existing one or not):
```
prepare()               process() *n               apply_config()          process() *n                   reset()
MaxxEffect_Message() -> MaxxEffect_Process() *n -> MaxxEffect_Message() -> MaxxEffect_Process() *n ... -> MaxxEffect_Reset()
```
If that can be handled properly (no risk for producing processing errors or sound distortion) then it's fine; otherwise, you may consider to de-couple "save-to-cache" and "apply-from-cache", i.e. only "save-to-cache" on handle_param_message() and "apply-from-cache" on prepare(), which prevents the interrupt while deferring the new config to the next prepare()

Reply:
Hi Johny,
It would not produce processing errors or sound distortion. The "apply-from-cache" will call "waves_effect_message()", it will let our tuner to hear the sound change immediately during audio playback.

Reply:
I see. It's fine as long as that is the intended behavior.